### ISSUE TITLE
MEAN.js 0.4.0 with token authentication?

### ISSUE THREAD

**Issue by ricardodovalle** [meanjs/mean#389]
I would like to use MEAN.js with token authentication, but I am a  Node.js beginner and I am porting a Rails app to learn.

I created a small project and add some features, the `'api/signin'` route is returning the token (`loginToken`) encoded with json web token (`jwt`), but I need help.

How to connect and keep Angular session a live using token?
How to add `express-jwt` and to protect all `/api` route keeping the project organized?
What is the correct way to store the secret? `(users.authentication.server.controller.js and strategies/local-token.js)`?
What is the correct way to user `signout`?

*Working on: https://github.com/ricardodovalle/meanjs-token-auth*

Thank you very much.

**Comment by sylvainlap** [comment]
Use jwt-simple to create your JSON Web Token.
Once your client get a token, you can save it client-side, thanks to the LocalStorage. And then, at each requests to the API, the client must send the token.
On the backend side, the token has to be check at each request to the API, thanks to an Express middleware.

You can read more about the workflow here : http://www.sitepoint.com/using-json-web-tokens-node-js/

Good luck ;)
Sylvain

**Comment by ricardodovalle** [comment]
Hi @sylvainlap,
I used jsonwebtoken instead jwt-simple, because it looks like more updated.
The project is creating the token and sending inside a json field (https://github.com/ricardodovalle/meanjs-token-auth)

I don't know how to do "...the token has to be check at each request to the API, thanks to an Express middleware."

Thanks

**Comment by sylvainlap** [comment]
Are you using Passport.js ? If yes, you should use this Strategy : https://github.com/jaredhanson/passport-http-bearer

"passport.authenticate" will be the Express Middleware that will check for the token.

**Comment by ricardodovalle** [comment]
Thanks @sylvainlap :clap: 
I will work and try to use.

**Comment by acoulon99** [comment]
Working on updating our repo for a meanjs token auth example. Hopefully it'll be useful to the community :)
https://github.com/castlewhitehall/meanjs-with-token-auth

**Comment by denzelwamburu** [comment]
I found this great Token-based AngularJS Authentication https://github.com/sahat/satellizer and works with any app perfectly.I prefer integrating this with the boilerplate but the issue is i can't seem to get the auth files in Meanjs.Any way to get around this?

**Comment by acoulon99** [comment]
I would be interested in that as well

**Comment by sylvainlap** [comment]
This lib looks great, but redondant with Passport.
If you want Token Auth, as I already said, use the passport-http-bearer strategy.

If you really want to use Satellizer, you have to integrate it client side (Angular), and then see here:
https://github.com/sahat/satellizer/tree/master/examples/server/node to integrate server side (build the /auth/* routes).

**Comment by britztopher** [comment]
If you want to attach the token to each request using Angular, you can use a request interceptor then attach it to the Authorization key in the request headers.  That way any request you make will have a token attached to it.  This is how loopback does it, and it works great.

**Comment by ilanbiala** [comment]
This is something that isn't necessary in MEAN.js, so I'm going to close this issue unless anyone has a compelling argument.

**Comment by lirantal** [comment]
@ilanbiala I actually find this PR very interesting. Tokens are very popular around mobile apps and provide the benefits of a stateless app. We can take a look at it once 0.4 is out, but of course it's not a requirement for 0.4 itself.

**Comment by ilanbiala** [comment]
It makes lots of sense for mobile apps, but it does add some extra complexity to the Angular code I think. Since we have no mobile app setup in MEAN.js, aren't we just adding complexity for most users?

**Comment by britztopher** [comment]
@ilanbiala i use mean.js for a mobile app with ionic and cardova and using a token based auth would be much easier to work with.  I think just because there is no mobile app setup doesnt mean people arent using mean.js for mobile.  Especially, with mobile first using html5 like bootstrap or whatever.

**Comment by ilanbiala** [comment]
@britztopher how much did you change to get a mobile app with MEAN.js working?

**Comment by britztopher** [comment]
Well i have a desktop and mobile app that use the same backend.  For example, the mean.js application angular side runs on desktop and on mobile web.  I then have another application that runs with the ionic framework, (just angular directives built for mobile application), this uses cordova to build and package up my mobile application into a android and ios runnable application.  So if you are asking if I am using the angular front-end that comes with mean.js as a mobile app then no I do not use it as a mobile application per se.

**Comment by ilanbiala** [comment]
So what I'm saying is that having MEAN.js support a mobile app packaging setup like Cordova means adding more code that for most users will just bloat the library. I think the best thing is if users like you write blog posts and create videos about modifying MEAN.js for these scenarios. MEAN.js was designed to be flexible, so we can incorporate these changes, but are 95% of users going to make use of the code that we add in?

**Comment by wesleyfsmith** [comment]
@ilanbiala I agree with what you are saying. I'm actually building an iconic app that will talk with meanjs, but I don't think this should go in the main framework.

It sounds to me more like it should be added as a module, similar to mean-seo. I thought features that are common requests, but not so common that they should be part of core meanjs should be added as additional modules (such as file upload, paginations, ect.) I also think generally going the module route as we add more and more features in is smarter from an ecosystem building perspective.

**Comment by jloveland** [comment]
I am also interested in this feature. I would like to leverage the backend services API and allow authentication from a mobile app. 

There seem to be a lot of benefits using token auth such as not needing to keep authentication state, serving front end via CDN, no CSRF, and persistent token store. 

Is there interest in adding this to the generator project?

here's a nice blog that discusses token auth:
http://code.tutsplus.com/tutorials/token-based-authentication-with-angularjs-nodejs--cms-22543

**Comment by britztopher** [comment]
Also, can this just replace the session based authentication/authorization model?  I dont know if passport has a way to pass back a token rather than store a session object, but this might be a good feature to replace the currently existing pattern.

**Comment by SOSANA** [comment]
Count me as apart of 95% that want this option. There are benefits mentioned above. In fact was hoping to do a PR request on generator for iconic integration in future along with material design :)

**Comment by ilanbiala** [comment]
@SOSANA adding in lots of design frameworks as options requires maintaining them, and right now that's somewhat because each one requires different templates, logic, and changes to keep up with the versions.

**Comment by lirantal** [comment]
@ilanbiala I can think of scenarios where users will choose MEAN.JS for it's server-side only due to it's simple, and good structure and will implement their own UI (whether it's another service like a mobile app or whatever). In this use-case it makes sense for MEAN.JS to support tokens.

I see there's enough interest and a lot of activity on this item so let's keep it open and get back to it after we're through with 0.4 :)

**Comment by SOSANA** [comment]
@ilanbiala  @lirantal Thanks for having an open mind and considering our arguments for these new features. I was wondering if you're still using trello for or road map? If so can we open it to the public to speed up the 4.0 transition.

**Comment by guillermo85** [comment]
Tokens auth +1 :+1: good work @ricardodovalle !

**Comment by karan-kapoor90** [comment]
+1

**Comment by lirantal** [comment]
Not using trello @SOSANA so let's just keep this thread open and get back to it later on when time allows.

BTW, is there a PR against 0.4 to enable the token authentication or no work on that has started yet?

**Comment by britztopher** [comment]
I was going to start here shortly, as I need to have this in place for my use case.  I plan to use passport's documented way, and then add an http request interceptor on angular side to attach token to header.  Also, use a response interceptor to stash the token in a SessionStorage object.  What do you all think?

**Comment by dotch** [comment]
sounds good! 
you can also take a look at [generator-angular-fullstack](https://github.com/DaftMonk/generator-angular-fullstack) which relies on token-auth.

**Comment by lirantal** [comment]
+1 for generator-angular-fullstack - could be a good reference

**Comment by britztopher** [comment]
@lirantal is there a reason why you labeled this ticket as a generator ticket?  I was under the impression that the token based approach would replace the session based one using the 0.4.0 branch.  My process was going to start by cloning the repo checkout out the 0.4.0 branch then working on replacing it there.

**Comment by lirantal** [comment]
I did? probably by mistake

**Comment by britztopher** [comment]
Ok, so I have decided to forgo the passport way and dive into using [express-jwt](https://github.com/auth0/express-jwt) with [jsonwebtoken](https://github.com/auth0/node-jsonwebtoken).  I was just reading this [article](https://auth0.com/blog/2014/01/07/angularjs-authentication-with-cookies-vs-token/), and found very straightforward and beneficial.  Also, I wont need to create an Oauth2orize server within our node app.

**Comment by dotch** [comment]
You might also look into using [satellizer](https://github.com/sahat/satellizer/) on the client and as meantioned before, check out their [node.js example](https://github.com/sahat/satellizer/tree/master/examples/server/node)

**Comment by ricardodovalle** [comment]
Another alternative to [satellizer](https://github.com/sahat/satellizer/) is [ng-token-auth](https://github.com/lynndylanhurley/ng-token-auth).

**Comment by dotch** [comment]
With satellizer we could do the oauth2 flow on the client side and get rid of passport on the server.
This would make the interoperability with native applications, which do their oauth2 flow clientside instead of in a webview, easier.

**Comment by britztopher** [comment]
So as an update: after looking at satallizer all yesterday and seeing how it would fit within our application's front and backend. I started yesterday to try to do with @dotch  was saying.  So far I've removed the authentication pieces for passport on the backend and now trying to wire up the frontend to make all the auth calls via satallizer.  I will continue to evolve this, and will post updates to progress.  Implementing satallizer seems pretty straightforward, however, I need to just dig into how we are doing things now to better understand current mean.js structures and concepts.  Once I do that then just creating a couple of my own classes to handle auth in the init section of the backend will go smoothly and cleanly.  First is understanding the current structure as the 0.4.0 branch is somewhat different.  

@lirantal I guess there is no kind of ticket or agile board for these kind of issues?

**Comment by ilanbiala** [comment]
@lirantal are you planning on having this ready for 0.4.0 or 0.4.x? I think having it for 0.4.0 would delay the release even more and we should add it in later and switch to it completely in 0.5.0 if we decide it is a better solution to authentication.

**Comment by britztopher** [comment]
@ilanbiala  I am doing this off the 0.4.0 branch knowing that it will not go into the 0.4.0 release.  I am just using the 0.4.0 branch, so I dont have to migrate my stuff from 0.3.x later.  Hopefully this will make it in the release after 0.4.0.

**Comment by ilanbiala** [comment]
@lirantal has anything been merged into 0.3.0 (master) that should be in 0.4.0 and we need to copy over? Is there any way to check. Also, shouldn't we just merge the branches and finish 0.4.0 as master in the next few weeks and then release it very shortly after?

**Comment by britztopher** [comment]
So I got local login working and storing the token in local storage.  Im looking how authorization is taken care of because we will need to query the user on each request to get the roles for each token.  Also, do we want to store the jwt token in the user model or create a `token` model that will store the user_id and roles associated with that token?  I see we are using the `acl` library for access control so I am right now reading up on how we are doing authorization.  Is there any mean.js documentation on how authorization is done.  If not, maybe there should be.

**Comment by ilanbiala** [comment]
We need to finish up most of the documentation as some of the stuff like user authorization is still being refactored in #343.

**Comment by britztopher** [comment]
@ilanbiala it looks like its just going to be a change to user schema structure.  Do you know if #343 is going to be added to the 0.4.0 branch soon?  That ticket just ended with you guys asking him to put it in there

**Comment by ilanbiala** [comment]
Yes, we are waiting on a PR from @sielay for the 0.4.0 branch and pending a final review it should be merged in for 0.4.0.

**Comment by denzelwamburu** [comment]
@britztopher do you have the codes online so that we can see your progress with satellizer?

**Comment by denzelwamburu** [comment]
Was figuring out maybe the sessions and tokens can be stored in Redis just cause of its security features.

**Comment by britztopher** [comment]
@denzelwamburu I can push my changes to my branch, however, it is very minimal at this point and only issues tokens to people who have account locally.  Once @sielay's PR is merged for new authorization process then I will start to work on authorization with auth tokens.

**Comment by vko-online** [comment]
Any update on this? +1 for token auth

**Comment by hougie** [comment]
+1 Token Auth

**Comment by britztopher** [comment]
@vko-online still waiting on @sielay's PR from #343 into 0.4.0 branch.

**Comment by sielay** [comment]
Now you wait for #559 ;) I had to create new PR to align it to 0.4.0. So try maybe to rebase and test against it. I had bit hectic time and need to add last fixes, then guys will take a look. Will try to add path asap.

**Comment by jloveland** [comment]
Any update on token auth? What's the plan now that #559 is stalled out?

**Comment by lirantal** [comment]
It's on the to-do list.

**Comment by rweindl** [comment]
Are there any new news on the progress of that topic?

**Comment by jloveland** [comment]
Is anyone actively working on this? @sielay? If not, is anyone out there wiling to take this on?

**Comment by mleanos** [comment]
I've been wanting this as well.. I'll make a run at it this week & submit a PR if I make any progress.

Any additional considerations, since this issue is old & the structure of the application has changed?

**Comment by britztopher** [comment]
@mleanos - if @sielay finished what he needed to do for user which will affect how authorization is handled then I was planning on getting this in.  I have done some work, mainly got satalizer to work with local token, but i need to know how authorization is going to be handled, (via the token, or by what we are doing now)

**Comment by mleanos** [comment]
@britztopher I think the PR from @sielay isn't going to be worked on anymore. I think this PR is taking it over... https://github.com/meanjs/mean/pull/752 but I'm not 100% sure.

**Comment by lirantal** [comment]
It has been on my to do list but I haven't gotten around to continue pursuing this yet except from general POC. How about we begin collaborating on a shared branch to tackle this? it shouldn't be that much work actually.

**Comment by mleanos** [comment]
@lirantal I think working on a shared branch is a good idea. I'm in :)

**Comment by britztopher** [comment]
Sure thing.  Lets open a new issue ticket to to get new conversation started to which way we should take this issue

**Comment by vko-online** [comment]
TL;DR Does 0.4.1 support Token auth? If not, why

**Comment by simison** [comment]
@vko-online See above https://github.com/meanjs/mean/issues/389#issuecomment-140300679

**Comment by vko-online** [comment]
@lirantal sure not much work, i have my app using token auth, and it really works fine, but specific to my app, anyway would be awesome to have it in this repo

**Comment by simison** [comment]
@vko-online nice. :-) https://github.com/vko-online/meanjs-token-auth

**Comment by mleanos** [comment]
@vko-online Your token auth implementation looks really good. A PR would be greatly appreciated. It would be nice to get a head start on this, so it will be ready for the 0.5.0.

**Comment by lirantal** [comment]
:+1:

**Comment by jloveland** [comment]
@lirantal, should this be assigned to milestone 0.5.x?

**Comment by kokokenada** [comment]
+1 Token Auth

**Comment by ilanbiala** [comment]
For anyone following this issue, this would be a + to have, but we already have a list of things we plan to add for 0.5. If you would like this feature, we'll gladly accept a PR.

**Comment by mleanos** [comment]
I realize that #1040 exists.. I've expressed my thoughts on this in the discussion in that PR, but I wanted to bring the discussion back over here since a lot has been discussed here.

A lot of good work went into @kokokenada's PR, and it has some good implementations that we can use. However, I think the general consensus is that we need a more focused approach with smaller commits/PR's.

I had already started on a new branch for Token Auth before #1040 was submitted. I thought it would be a shame to not take advantage of the work that was put into this by @kokokenada & myself. 

@lirantal You mentioned that we might want to work on this together in a shared branch. I already proposed this in the above mentioned PR. 

I think we should start working together through my branch. It has smaller commits that have a bit more focus. It should be rather easy to collaborate on it. I originally posted a link to a branch where I started this implementation, but I ended up creating a fresh branch; I learned quite a bit with my first run and feel like it would be better to start fresh.

https://github.com/mleanos/mean/commits/token-auth-simplified

The commit messages explain what I have done so far. Basically, the back-end is done. I haven't touched the front-end at all.

Anyone have thoughts on my proposed approach?

**Comment by kokokenada** [comment]
Great approach and much appreciated.  I will try working with your branch as soon as I am able .

**Comment by mleanos** [comment]
@kokokenada Great, thank you! I was thinking that you might want to take a shot at the front-end implementation. 

Earlier in this discussion, there was some talk of integrating Satellizer on the front-end for the Token management. It looks like a great library, and I think a bit of the struggles that come with this type of implementation are taken care of for you. Do you want to take a look at it?

https://github.com/sahat/satellizer

I don't think we need to add anything to the back-end for this. Since we already have the OAuth routes setup & working. However, I may be wrong on that.

**Comment by kokokenada** [comment]
Satelizer looks great. How do the leaders/gatekeepers feel about it as a dependency for MEANJS? (Who are the leaders/gatekeepers?) It isn't clear to me how Satelizer relates to the large amounts of strategy providers already in MEANJS. I'd have a lot learning to do if I were to take it on as I don't have experience in the area.  I'm heads down on some other commitments for a couple of weeks at least.  (One of which is building out a MEANJS development team in Toronto if that strikes any chords with anyone.)  I'm happy to help, but I may not be the best lead for this important work.  Definitely comfy in a supporting role.

**Comment by mleanos** [comment]
Anybody want to start working on this together?

I have a preliminary attempt at this on the server-side here, https://github.com/mleanos/mean/commits/token-auth-simplified

@lirantal Any thoughts on organizing the effort here?

**Comment by lirantal** [comment]
anyone who has time could just get onboard, I don't think any formal organization is required.
unfortunately I don't have much time to invest in this at the moment, but goodluck! :)

**Comment by mleanos** [comment]
Added https://github.com/meanjs/mean/pull/1121 to get feedback & track

Anyone that wants to help out, feel free to submit changes to my fork's branch.

**Comment by trainerbill** [comment]
I think we can close this out.  Feature is done and awaiting feedback/merge:

https://github.com/meanjs/mean/pull/1163

Please comment to get it pushed through

**Comment by codydaig** [comment]
@trainerbill If your commit message in line with the Contributing.md document, this issue will auto close when #1163 is merged.

**Comment by trendzetter** [comment]
I am installing this now. I browsed the code and saw that there is a default secret. Not a huge issue for now probably but maybe something to avoid? I don't know what is the best solution. It's easy to update but at some point lazy people will probably install it like that in production.

**Comment by simison** [comment]
@trendzetter https://github.com/meanjs/mean/blob/43c7041aa574820e0854f095ec96e8ea3b535053/config/config.js#L90-L110 ;-)

**Comment by trendzetter** [comment]
so it refuses to run in production unless you add your own secret. Nice!

**Comment by trendzetter** [comment]
I am mistaken. It's merely warning but not refusing to run. 
So far no further issues found.

**Comment by lirantal** [comment]
@trainerbill I'll close it when we get the JWT PR closed too :)

**Comment by leafdropco** [comment]
I think it would be extremely useful to integrate mobile features. Maybe create a separate build specifically for mobile? I'm new to the mean stack and have been building my first big application. After reading the thread, I'm a little bummed that MEAN.js isn't as mobile friendly as I anticipated.

**Comment by denzelwamburu** [comment]
Can you elaborate more on mobile features?

**Comment by leafdropco** [comment]
I was referring to some of the previous suggestions like JWT.
Maybe even adding Material Design instead of bootstrap, I think that would entice people to use meanjs over ionic or mean.io. Especially if you could integrate cordova.

**Comment by vko-online** [comment]
So, for now, we don't have TokenAuth, ES2015, MaterialDesign, Swagger

Also would be nice to have promise based Mongoose api in generator

```
exports.update = (req, res) => {
  var order = req.order;
  order = Object.assign(order, req.body); // <== lodash is redundant
  order.save()
    .then(() => res.jsonp(order))
    .catch(err => res.status(400).send({ message: errorHandler.getErrorMessage(err) }));
};
```
looks much nicer than
```
exports.update = (req, res) => {
  var order = req.order;
  order = _.extend(order, req.body);
  order.save(function (err, order){
    if(err){
      return res.status(400).send({ message: errorHandler.getErrorMessage(err) });
    } else {
      res.jsonp(order);
    }
  });
};

```
